kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java

```diff
@@ -0,0 +1,225 @@
+import org.springframework.integration.kafka.core.Partition;
+import org.springframework.integration.kafka.core.ZookeeperConfiguration;
+import org.springframework.integration.kafka.inbound.KafkaMessageDrivenChannelAdapter;
+import org.springframework.integration.kafka.listener.KafkaMessageListenerContainer;
+import org.springframework.integration.kafka.listener.KafkaNativeOffsetManager;
+import org.springframework.integration.kafka.listener.OffsetManager;
+import org.springframework.integration.kafka.support.ZookeeperConnect;
+import org.springframework.retry.RetryCallback;
+import org.springframework.retry.RetryContext;
+import org.springframework.retry.backoff.ExponentialBackOffPolicy;
+import org.springframework.retry.policy.SimpleRetryPolicy;
+import org.springframework.retry.support.RetryTemplate;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.ObjectUtils;
+import org.springframework.util.StringUtils;
+
+import kafka.serializer.Decoder;
+import kafka.serializer.DefaultDecoder;
+import kafka.utils.VerifiableProperties;
+
+/**
+ * Source module that receives data from Kafka.
+ *
+ * @author Soby Chacko
+ */
+@EnableBinding(Source.class)
+@EnableConfigurationProperties(KafkaConfigurationProperties.class)
+public class KafkaSource {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private static final int METADATA_VERIFICATION_RETRY_ATTEMPTS = 10;
+
+    private static final double METADATA_VERIFICATION_RETRY_BACKOFF_MULTIPLIER = 2;
+
+    private static final int METADATA_VERIFICATION_RETRY_INITIAL_INTERVAL = 100;
+
+    private static final int METADATA_VERIFICATION_MAX_INTERVAL = 1000;
+
+    @Autowired
+    @Bindings(KafkaSource.class)
+    private Source source;
+
+    @Autowired
+    private KafkaConfigurationProperties properties;
+
+    @Bean
+    public KafkaMessageListenerContainer container() {
+        Partition[] partitions = getPartitions();
+
+        KafkaMessageListenerContainer container = ObjectUtils.isEmpty(partitions) ?
+                new KafkaMessageListenerContainer(kafkaConnectionFactory(), properties.getTopics()) :
+                new KafkaMessageListenerContainer(kafkaConnectionFactory(), partitions);
+
+        // if we have less target partitions than target concurrency, adjust accordingly
+        container.setConcurrency(Math.min(properties.getConcurrency(), partitions.length));
+
+        container.setMaxFetch(properties.getMaxFetch());
+        container.setStopTimeout(properties.getStopTimeout());
+        container.setQueueSize(properties.getQueueSize());
+        container.setOffsetManager(kafkaNativeOffsetManager(kafkaConnectionFactory()));
+
+        return container;
+    }
+
+    @Bean
+    public KafkaMessageDrivenChannelAdapter kafkaMessageDrivenChannelAdapter() throws IllegalAccessException, InstantiationException, ClassNotFoundException {
+        final KafkaMessageDrivenChannelAdapter kafkaMessageDrivenChannelAdapter =
```

Comment:
Since you don't use SI Java DSL anymore, the `spring-integration-java-dsl` dependency can be removed.